service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java

```diff
@@ -482,4 +506,67 @@ public void registerLivenessChecks(final HealthCheckHandler handler) {
         });
     }
+
+    private Future<JsonObject> getCredentialsForDevice(final String tenantId, final String type, final String authId) {
+
+        Future<JsonObject> result = Future.future();
+        getCredentialsClient(tenantId).compose(client -> {
+            Future<CredentialsResult> credResultFuture = Future.future();
+            client.get(type, authId, credResultFuture.completer());
+            return credResultFuture;
+        }).compose(credResult -> {
+            if (credResult.getStatus() == HttpURLConnection.HTTP_OK) {
+                JsonObject payload = credResult.getPayload();
+                result.complete(payload);
+            } else if (credResult.getStatus() == HttpURLConnection.HTTP_NOT_FOUND) {
+                result.fail(String.format("cannot retrieve credentials (not found for type <%s>, authId <%s>)",type,authId));
+            } else {
+                result.fail("cannot retrieve credentials");
+            }
+        }, result);
+
+        return result;
+    }
+
+    /**
+     * Validates an authentication object against credentials secrets available by the get operation of the
+     *  <a href="https://www.eclipse.org/hono/api/Credentials-API/">Credentials API</a>.
+     * <p>The credentials are first retrieved from the credentials service, and then all matching validators are
+     * invoked until one is successful or all failed. The authentication object is validated iff at least
+     * one validator was successful.
+     *
+     * @param tenantId The tenantId to which the device belongs.
+     * @param type The type of credentials that are to be used for validation.
+     * @param authId The authId of the credentials that are to be used for validation.
+     * @param authenticationObject The authentication object to be validated, e.g. a password, a preshared-key, etc.
+
+     * @return Future The future object carrying the payload of the credentials get operation, if successful.
+     */
+    protected Future<JsonObject> validateCredentialsForDevice(final String tenantId, final String type, final String authId,
+                                                              final Object authenticationObject) {
+        Future<JsonObject> credResult = Future.future();
+
+        credResult = getCredentialsForDevice(tenantId, type, authId).compose(payload -> {
+            Future<JsonObject> result = Future.future();
+            Set<CredentialsSecretsValidator> validators = getCredentialsUtils().findAppropriateValidators(type);
+            if (validators == null || validators.size() == 0) {
+                result.fail(String.format("Cannot find validators for type <%s>", type));
+                return result;
+            }
+
+            // since several validators may exist per type, they are invoked until one is successful (or none)
+            Predicate<CredentialsSecretsValidator> validationPred =
+                    validator -> validator.validate(payload, authenticationObject);
```

Comment:
this looks like it will produce a `ClassCastException` if *authenticationObject* is of a different type than the parameter class of the validators ...